lib/Dialect/FIRRTL/Transforms/IMDeadCodeElim.cpp

```diff
@@ -399,6 +406,46 @@ void IMDeadCodeElimPass::rewriteModuleSignature(FModuleOp module) {
   numRemovedPorts += deadPortIndexes.size();
 }
 
+void IMDeadCodeElimPass::eraseEmptyModule(FModuleOp module) {
+  // Public modules cannot be erased.
+  if (module.isPublic())
+    return;
+
+  // If the module doesn't have arguments, operations or annotations, we
+  // consider it to be dead.
+  if (!module.getBody()->args_empty() || !module.getBody()->empty() ||
+      !module.annotations().empty())
+    return;
+
+  // If there is "output_file" attribute, we don't delete the module.
+  if (module->hasAttr("output_file"))
+    return;
+
+  // Ok, the module is empty. Delete instances unless they have symbols.
+  LLVM_DEBUG(llvm::dbgs() << "Erase " << module.getName() << "\n");
+
+  InstanceGraphNode *instanceGraphNode =
+      instanceGraph->lookup(module.moduleNameAttr());
+
+  bool existsInstanceWithSymbol = false;
```

Comment:
Another option would be to delete instances during the normal processing, then modules would be trivially dead.  I think deadness of an instance is no args and points to an empty module.

Reply:
> Another option would be to delete instances during the normal processing, then modules would be trivially dead. I think deadness of an instance is no args and points to an empty module.

Currently normal processing is executed parallelly so it is not possible to check emptiness of other modules there. Also dead modules might contain  instances of other dead module. Therefore, it is necessary to delete instances in the separate procedure.

Reply:
With all the changes can we delete the instances while processing the modules in parallel?  I'm hoping we can figure out a way to just check the `liveModule` set here and just delete the module without processing the body.